### PR TITLE
Fix multi-field unique values renderer when parsing AFS settings & basic label expression parsing

### PR DESCRIPTION
**Pull request by nirvn**
## Description
I swear the upgrade efforts end here, can't find anything else wrong with my AFS severs ;)

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/providers/arcgisrest/qgsarcgisrestutils.cpp [hunk 1]
```diff
@@ -919,8 +918,27 @@ QgsFeatureRenderer *QgsArcGisRestUtils::parseEsriRenderer( const QVariantMap &re
...
+    const QString field1 = rendererData.value( QStringLiteral( "field1" ) ).toString();
+    const QString field2 = rendererData.value( QStringLiteral( "field2" ) ).toString();
+    const QString field3 = rendererData.value( QStringLiteral( "field3" ) ).toString();
+    QString attribute;
+    if ( !field2.isEmpty() || !field3.isEmpty() )
+    {
+      const QString delimiter = rendererData.value( QStringLiteral( "fieldDelimiter" ) ).toString();
+      if ( !field3.isEmpty() )
+      {
+        attribute = QString( "concat(\"%1\",'%2',\"%3\",'%4',\"%5\")" ).arg( field1, delimiter, field2, delimiter, field3 );
```
Comment (nyalldawson) [inline src/providers/arcgisrest/qgsarcgisrestutils.cpp]:
```suggestion
        attribute = QStringLiteral( "concat(\"%1\",'%2',\"%3\",'%4',\"%5\")" ).arg( field1, delimiter, field2, delimiter, field3 );
```

File: src/providers/arcgisrest/qgsarcgisrestutils.cpp [hunk 2]
```diff
@@ -919,8 +918,27 @@ QgsFeatureRenderer *QgsArcGisRestUtils::parseEsriRenderer( const QVariantMap &re
...
+      }
+      else
+      {
+        attribute = QString( "concat(\"%1\",'%2',\"%3\")" ).arg( field1, delimiter, field2 );
```
Comment (nyalldawson) [inline src/providers/arcgisrest/qgsarcgisrestutils.cpp]:
```suggestion
        attribute = QStringLiteral( "concat(\"%1\",'%2',\"%3\")" ).arg( field1, delimiter, field2 );
```

File: src/providers/arcgisrest/qgsarcgisrestutils.cpp [hunk 3]
```diff
@@ -964,6 +982,24 @@ QgsFeatureRenderer *QgsArcGisRestUtils::parseEsriRenderer( const QVariantMap &re
   return nullptr;
 }
 
+QString QgsArcGisRestUtils::parseEsriArcadeExpression( const QString &string )
+{
+  QString expression = string;
+
+  // Replace a few Arcade token to QGIS equivalents
+  expression = expression.replace( QRegularExpression( "(?=([^\"\\\\]*(\\\\.|\"([^\"\\\\]*\\\\.)*[^\"\\\\]*\"))*[^\"]*$)(\\s|^)CONCAT(\\s|$)" ), QString( "\\4||\\5" ) );
```
Comment (nyalldawson) [inline src/providers/arcgisrest/qgsarcgisrestutils.cpp]:
```suggestion
  expression = expression.replace( QRegularExpression( "(?=([^\"\\\\]*(\\\\.|\"([^\"\\\\]*\\\\.)*[^\"\\\\]*\"))*[^\"]*$)(\\s|^)CONCAT(\\s|$)" ), QStringLiteral( "\\4||\\5" ) );
```

File: src/providers/arcgisrest/qgsarcgisrestutils.cpp [hunk 4]
```diff
@@ -964,6 +982,24 @@ QgsFeatureRenderer *QgsArcGisRestUtils::parseEsriRenderer( const QVariantMap &re
...
+  expression = expression.replace( QRegularExpression( "(?=([^\"\\\\]*(\\\\.|\"([^\"\\\\]*\\\\.)*[^\"\\\\]*\"))*[^\"]*$)(\\s|^)NEWLINE(\\s|$)" ), QString( "\\4'\\n'\\5" ) );
```
Comment (nyalldawson) [inline src/providers/arcgisrest/qgsarcgisrestutils.cpp]: etc ;)

File: src/providers/arcgisrest/qgsarcgisrestutils.cpp [hunk 5]
```diff
@@ -964,6 +982,24 @@ QgsFeatureRenderer *QgsArcGisRestUtils::parseEsriRenderer( const QVariantMap &re
   return nullptr;
 }
 
+QString QgsArcGisRestUtils::parseEsriArcadeExpression( const QString &string )
```
Comment (nyalldawson) [inline src/providers/arcgisrest/qgsarcgisrestutils.cpp]: Better add some unit tests for this method -- it looks complex, and is likely only going to get more so!

Comment (nirvn) [reply src/providers/arcgisrest/qgsarcgisrestutils.cpp]: @nyalldawson , the testParseLabeling test covers all that's currently implemented in parseEsriArcadeExpression, is that satisfactory for now?

File: src/providers/arcgisrest/qgsarcgisrestutils.cpp (hunk 4 above)
Comment (nirvn) [reply src/providers/arcgisrest/qgsarcgisrestutils.cpp]: @nyalldawson , yeah, indeed, etc. :smile: I'm not in need of anything else ATM, I'm sure others will eventually expand.

File: src/providers/arcgisrest/qgsarcgisrestutils.cpp [hunk 6]
```diff
@@ -964,6 +982,24 @@ QgsFeatureRenderer *QgsArcGisRestUtils::parseEsriRenderer( const QVariantMap &re
   return nullptr;
 }
 
+QString QgsArcGisRestUtils::parseEsriLabelingExpression( const QString &string )
+{
+  QString expression = string;
+
+  // Replace a few ArcGIS token to QGIS equivalents
+  expression = expression.replace( QRegularExpression( "(?=([^\"\\\\]*(\\\\.|\"([^\"\\\\]*\\\\.)*[^\"\\\\]*\"))*[^\"]*$)(\\s|^)CONCAT(\\s|$)" ), QStringLiteral( "\\4||\\5" ) );
```
Comment (elpaso) [inline src/providers/arcgisrest/qgsarcgisrestutils.cpp]: @nirvn  Sorry to jump in late, just a minor recommendation: in order to make the regexp more readable you could use the raw string syntax and skip all that \\ escaping, something like this:
```c++
expression = expression.replace( QRegularExpression( R"re((?=([^"\\]*(\.|"([^"\\]*\.)*[^"\]*"))*[^"]*$)(\s|^)CONCAT(\s|$))re" ), QStringLiteral( R"re(\4||\5)re" ) );
```

**Pull request merged**